### PR TITLE
#498 should add support for extension dependency validation

### PR DESCRIPTION
**Pull request by phuonglm**
Connect #498

### CODE REVIEW

File: lib/teracy-dev/extension/manager.rb [hunk 1]
```diff
@@ -64,6 +65,32 @@ def validate(extension)
...
       end
 
+      def validate_dependencies(extension, extensions)
+        extension_manifest = Manager.manifest(extension)
+        return unless Util.exist?(extension_manifest['dependencies'])
+
+        extension_manifest['dependencies'].each do |dependency|
+          found = nil
+          extensions.each do |target_extension|
+            target_manifest = Manager.manifest(target_extension)
```
Comment (datphan) [inline lib/teracy-dev/extension/manager.rb]: I think this is too many loop, consider trying this:
```
extention_manifest_list = extensions.map { |x| Manager.manifest(x) }

extension_manifest['dependencies'].each do |dependency|
  found = extention_manifest_list.find { |x| x['name'] == dependency['name]}
...
```

File: lib/teracy-dev/extension/manager.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ def install(extensions)
         extensions.each do |extension|
           sync(extension)
           validate(extension)
+          validate_dependencies(extension, extensions)
```
Comment (datphan) [inline lib/teracy-dev/extension/manager.rb]: should validate dependencies if extension is enabled, so it should move to `validate()`

File: lib/teracy-dev/extension/manager.rb [hunk 3]
```diff
@@ -20,17 +20,24 @@ def install(extensions)
...
+            sync(extension)
+            validate(extension)
+          end
         end
+
+        extention_manifest_list = extensions.map { |x| Manager.manifest(x) if Util.true?(x['enabled'])}.compact
+        extensions.each do |extension|
+          validate_dependencies(extension, extention_manifest_list) if Util.true?(extension['enabled'])
+        end
+
```
Comment (datphan) [inline lib/teracy-dev/extension/manager.rb:33]: we should loop this 1 time:
```
extention_manifest_list = ..
extensions.each do |extension|
    sync(extension)
    validate(extension, extention_manifest_list)
end
```

File: lib/teracy-dev/extension/manager.rb [hunk 4]
```diff
@@ -20,17 +20,24 @@ def install(extensions)
...
         timer_end = Time.now
         @logger.debug("installation finished in #{timer_end - timer_start}s of extensions: #{extensions}")
       end
 
       private
 
       def sync(extension)
-        return unless Util.true?(extension['enabled'])
```
Comment (datphan) [inline lib/teracy-dev/extension/manager.rb]: `Location::Manager.sync` is also using by others, so should have this line

File: lib/teracy-dev/extension/manager.rb [hunk 5]
```diff
@@ -42,7 +49,6 @@ def sync(extension)
       end
 
       def validate(extension)
-        return unless Util.true?(extension['enabled'])
```
Comment (datphan) [inline lib/teracy-dev/extension/manager.rb]: should be the same logic above, so should not remove this line

Comment (datphan) [reply lib/teracy-dev/extension/manager.rb]: we should move `validate_dependencies` at bottom of this function

File: lib/teracy-dev/extension/manager.rb (hunk 3 above)
Comment (phuonglm) [reply lib/teracy-dev/extension/manager.rb:33]: We need to finish sync all extension before do version check. For example i require V5 but current on local repo is V4 remote repo is v5. manifest list must load after sync too

**Comment by phuonglm** [comment]
I updated pr please review

**Review by datphan — APPROVED** [review]

**Review by hoatle — APPROVED** [review]
LGTM

**Pull request merged**